src/client/Microsoft.Identity.Client/OAuth2/TokenClient.cs

```diff
@@ -169,31 +167,55 @@ private void AddBodyParamsAndHeaders(IDictionary<string, string> additionalBodyP
                     _serviceBundle.HttpTelemetryManager.GetLastRequestHeader());
             }
 
-            //Signaling that the client can perform PKey Auth on supported platforms
+            // Signaling that the client can perform PKey Auth on supported platforms
             if (DeviceAuthHelper.CanOSPerformPKeyAuth())
             {
                 _oAuth2Client.AddHeader(PKeyAuthConstants.DeviceAuthHeaderName, PKeyAuthConstants.DeviceAuthHeaderValue);
             }
 
-            AddExtraHttpHeaders();
+            AddCcsHeader(additionalBodyParameters);
         }
 
-        private void AddExtraHttpHeaders()
+        // The CCS header is used by the CCS service to help route requests to resources in Azure to speed up authentication.
+        // It consists of either the {ObjectId}@{TenantId} or the upn of the current user.
+        // See https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2525
+        private void AddCcsHeader(IDictionary<string, string> additionalBodyParameters)
         {
-            if (_requestParams.ExtraHttpHeaders != null)
+            string ccsHeaderHint = string.Empty;
+            if (!string.IsNullOrEmpty(_requestParams.CcsRoutingHint))
+            {
+                ccsHeaderHint = $"oid:{_requestParams.CcsRoutingHint}";
+            }
+
+            else if(_requestParams?.Account?.HomeAccountId != null)
             {
-                foreach (KeyValuePair<string, string> pair in _requestParams.ExtraHttpHeaders)
+                if (!string.IsNullOrEmpty(_requestParams.Account.HomeAccountId.Identifier))
```

Comment:
Perhaps these two `ifs` can be combined into:
`else if (!string.IsNullOrEmpty(_requestParams?.Account?.HomeAccountId?.Identifier))`